Improve the precision of abs() and sign() for large values

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99550

@ev-br found in
https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/117#issuecomment-1514959633
that the precision of `abs()` for large values in the vectorised case is less-than-good.
This PR fixes this issue. While doing that, we are able to comment out a
few tests on extremal values.

Fixes https://github.com/pytorch/pytorch/issues/53958 https://github.com/pytorch/pytorch/issues/48486

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

lezcano committed:
Improve the precision of abs() and sign() for large values

@ev-br found in
https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/117#issuecomment-1514959633
that the precision of `abs()` for large values in the vectorised case is less-than-good.
This PR fixes this issue. While doing that, we are able to comment out a
few tests on extremal values.

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/cpu/vec/vec256/vec256_complex_double.h:
```diff
@@ -140,14 +135,15 @@ template <> class Vectorized<c10::complex<double>> {
   Vectorized<c10::complex<double>> sgn() const {
-    auto abs = abs_();
+    auto abs_zero = abs().values;             // abs 0
+    auto abs = _mm256_movedup_pd(abs_zero);  // abs abs
```
You can reduce the number of mask and shuffle ops by just changing `abs_` to:
```
  __m256d abs_() const {
    auto real = _mm256_movedup_pd(values);  // real real
    auto imag = _mm256_movehdup_pd(values); // imag imag
    return Sleef_hypotd4_u05(real, imag);   // abs  abs
  }
```

peterbell10 replied:
and with this change only `abs_()` has to be modified.

lezcano replied:
Oh, great point! I'll do that.

lezcano committed:
Update on "Improve the precision of abs() and sign() for large values"


ev-br found in
https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/117#issuecomment-1514959633
that the precision of `abs()` for large values in the vectorised case is less-than-good.
This PR fixes this issue. While doing that, we are able to comment out a
few tests on extremal values.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
@peterbell10 addressed the review. I also cleaned `sgn` a bit, making it more uniform across the 4 implementations.

lezcano committed:
Update on "Improve the precision of abs() and sign() for large values"


ev-br found in
https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/117#issuecomment-1514959633
that the precision of `abs()` for large values in the vectorised case is less-than-good.
This PR fixes this issue. While doing that, we are able to comment out a
few tests on extremal values.

Fixes https://github.com/pytorch/pytorch/issues/53958 https://github.com/pytorch/pytorch/issues/48486

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

lezcano commented:
@pytorch merge

lezcano commented:
@pytorchbot merge

Flamefire commented:
Unfortunately the VSX code hasn't been updated and produces NaNs where inf is expected for those test_reference_numerics_extremal__refs_abs_cpu_complex* tests

lezcano commented:
Feel free to send a PR

Flamefire commented:
Done: https://github.com/pytorch/pytorch/pull/116859
